### PR TITLE
Python timing APIs

### PR DESCRIPTION
**Pull request by VidyaChhabria**
Added API support for the following pin properties:
1. Arrival
2. Slew
3. Endpoints
Startpoint is currently TBD. 

Cleaned up the code based on the comments from the previous pull requests and removed string-based arguments. 
Also refactored the code to move the timing-related APIs to a separate timing class for better API management.

### CODE REVIEW

File: src/Timing.cc [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+  return isStartpoint_(sta_pin);
+}
+
+bool Timing::isStartpoint(odb::dbBTerm* db_pin)
+{
+  sta::Pin* sta_pin = getSta()->getDbNetwork()->dbToSta(db_pin);
+  return isStartpoint_(sta_pin);
+}
+
+bool Timing::isStartpoint_(sta::Pin* sta_pin)
```
[the remark this answers is not in the document]
Comment (VidyaChhabria) [reply src/Timing.cc:111]: @maliberty I am not sure how to implement this API. Do you have any suggestions on this?  I cleaned up the rest of the APIs and refractored all the code.

**Comment by maliberty** [comment]
> Cleaned up the code based on the comments from the previous pull requests and removed string-based arguments.

Does this mean #4326 should be closed?

**Comment by VidyaChhabria** [comment]
> > Cleaned up the code based on the comments from the previous pull requests and removed string-based arguments.
> 
> Does this mean #4326 should be closed?

Yes we can close #4443.

File: src/Design.cc [hunk 2]
```diff
@@ -266,6 +257,26 @@ bool Design::isInClock(odb::dbInst* inst)
...
+{
+  auto* mTerm = iterm->getMTerm();
+  if (mTerm != nullptr) {
+    auto sig_type = mTerm->getSigType();
+    if (sig_type == odb::dbSigType::POWER
+        || sig_type == odb::dbSigType::GROUND) {
+      return true;
+    }
+  }
+  return false;
```
Comment (maliberty) [inline src/Design.cc:277]: Just
````
return iterm->getSigType().isSupply();
```

File: include/ord/Design.h [hunk 3]
```diff
@@ -138,6 +139,8 @@ namespace ord {
 
 class Tech;
 
+class OpenRoad;
```
Comment (maliberty) [inline include/ord/Design.h:142]: Why is this needed?

File: include/ord/Timing.h [hunk 4]
```diff
@@ -0,0 +1,126 @@
+/////////////////////////////////////////////////////////////////////////////
+//
+// Copyright (c) 2022, The Regents of the University of California
```
Comment (maliberty) [inline include/ord/Timing.h:3]: 2023

File: include/ord/Timing.h [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+#include "sta/PatternMatch.hh"
+#include "sta/Sdc.hh"
+#include "sta/SdcClass.hh"
+
+namespace odb {
+// class dbBlock;
+class dbMaster;
+class dbMTerm;
+// class dbNet;
+// class dbInst;
```
Comment (maliberty) [inline include/ord/Timing.h:51]: rm dead code

File: src/Timing.cc [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+
+bool Timing::isEndpoint(odb::dbBTerm* db_pin)
+{
+  sta::Pin* sta_pin = getSta()->getDbNetwork()->dbToSta(db_pin);
+  return isEndpoint_(sta_pin);
+}
+
+bool Timing::isEndpoint_(sta::Pin* sta_pin)
+{
+  auto search = sta::Sta::sta()->search();
```
Comment (maliberty) [inline src/Timing.cc:89]: Use getSta() instead of accessing this singleton (throughout).  dbSta inherits from  Sta.

File: src/Timing.cc [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+
+float Timing::slew_corner(sta::Vertex* vertex)
+{
+  sta::Sta* sta = sta::Sta::sta();
+  float slew_max = -sta::INF;
+  for (auto corner : getCorners()) {
+    slew_max = std::max(
+        slew_max,
+        sta::delayAsFloat(sta->vertexSlew(
+            vertex, sta::RiseFall::rise(), corner, sta::MinMax::max())));
```
Comment (maliberty) [inline src/Timing.cc:132]: do you always want max() ?

File: src/Timing.cc [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+  // auto vertex_array = vertices(sta_pin);
+  // for (auto vertex : vertex_array) {
+  //  //    if (vertex != nullptr && search->isStartpoint(vertex) {
+  //  //      return true;
+  //  //    }
+  //}
+  return false;
+}
+
+float Timing::slew_corner(sta::Vertex* vertex)
```
Comment (maliberty) [inline src/Timing.cc:124]: This is an odd name as no corner is involved in the API

File: include/ord/Timing.h [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+  float getPinArrivalTime(sta::Clock* clk,
+                          const sta::RiseFall* clk_rf,
+                          sta::Vertex* vertex,
+                          const sta::RiseFall* arrive_or_hold);
+  sta::Graph* cmdGraph();
+  sta::Network* cmdLinkedNetwork();
+  std::array<sta::Vertex*, 2> vertices(const sta::Pin* pin);
+  bool isTimeInf(float time);
+
+  float slew_corner(sta::Vertex* vertex);
```
Comment (maliberty) [inline include/ord/Timing.h:102]: Why do we need to expose users to Vertex or Pin?  Can these API be written in terms of odb objects alone?

File: src/Timing.cc [hunk 10]
```diff
@@ -0,0 +1,334 @@
...
+      pinSlew = std::max(pinSlew, pinSlewTemp);
+    }
+  }
+  return pinSlew;
+}
+
+sta::Network* Timing::cmdLinkedNetwork()
+{
+  sta::Network* network = sta::Sta::sta()->cmdNetwork();
+  ;
```
Comment (maliberty) [inline src/Timing.cc:167]: rm

File: include/ord/Timing.h [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+  };
+
+  sta::ClockSeq findClocksMatching(const char* pattern,
+                                   bool regexp,
+                                   bool nocase);
+  float getPinArrival(odb::dbITerm* db_pin, RiseFall rf);
+  float getPinArrival(odb::dbBTerm* db_pin, RiseFall rf);
+  std::vector<float> arrivalsClk(const sta::RiseFall* rf,
+                                 sta::Clock* clk,
+                                 const sta::RiseFall* clk_rf,
```
Comment (maliberty) [inline include/ord/Timing.h:91]: You have defined RiseFall here - why also use sta::RiseFall?  It seems confusing

**Review by maliberty — CHANGES_REQUESTED** [review]

File: src/Timing.cc (hunk 7 above)
Comment (VidyaChhabria) [reply src/Timing.cc]: I've added an option to pick between min and max for the slew and arrival times.

File: include/ord/Timing.h (hunk 9 above)
Comment (VidyaChhabria) [reply include/ord/Timing.h]: Yes, that makes sense. I'll  move references to internal STA objects to private.

**Review by maliberty — APPROVED** [review]
